mycpp/gc_str.cc

```diff
@@ -462,3 +468,105 @@ List<Str*>* Str::split(Str* sep) {
 
   return result;
 }
+
+static inline Str* _StrFormat(const char* fmt, va_list args) {
+  int n = strlen(fmt);
+  auto beg = std::cregex_iterator(fmt, fmt + n, gStrFmtRegex);
+  auto end = std::cregex_iterator();
+
+  char int_buf[kMaxFmtWidth];
+  std::string buf;
+  for (std::cregex_iterator it = beg; it != end; ++it) {
+    const std::cmatch& match = *it;
+
+    const std::csub_match& lit_m = match[1];
+    assert(lit_m.matched);
+    const std::string& lit_s = lit_m.str();
+    buf.append(lit_s);
+
+    int width = 0;
+    bool zero_pad = false;
+    const std::csub_match& width_m = match[2];
+    const std::string& width_s = width_m.str();
+    if (width_m.matched && !width_s.empty()) {
+      if (width_s[0] == '0') {
+        zero_pad = true;
+        assert(width_s.size() > 1);
+        assert(StringToInteger(width_s.c_str() + 1, width_s.size() - 1, 10,
+                               &width));
+      } else {
+        assert(StringToInteger(width_s.c_str(), width_s.size(), 10, &width));
+      }
+      assert(width >= 0 && width < kMaxFmtWidth);
+    }
+
+    char const* str_to_add = nullptr;
+    const std::csub_match& code_m = match[3];
+    const std::string& code_s = code_m.str();
+    if (!code_m.matched) {
+      assert(!width_m.matched);  // python errors on invalid format operators
+      goto done;
+    }
+    assert(code_s.size() == 1);
+    switch (code_s[0]) {
+    case '%': {
+      assert(code_s.size() == 1);
+      str_to_add = code_s.c_str();
+      break;
+    }
+    case 's': {
+      Str* s = va_arg(args, Str*);
+      str_to_add = s->data();
```

Comment:
Hm technically this does not follow Python semantics, where NULs aren't treated as special, unlike in C

```
>>> s = '\0 foo'
>>> 'format %s' % s
'format \x00 foo'
```

Reply:
ah great catch. i'll add a test for this